src/wallet/dump.cpp

```diff
@@ -87,3 +87,171 @@ bool DumpWallet(std::shared_ptr<CWallet> wallet, bilingual_str& error)
 
     return ret;
 }
+
+// The standard wallet deleter function blocks on the validation interface
+// queue, which doesn't exist for the bitcoin-wallet. Define our own
+// deleter here.
+static void WalletToolReleaseWallet(CWallet* wallet)
+{
+    wallet->WalletLogPrintf("Releasing wallet\n");
+    wallet->Close();
+    delete wallet;
+}
+
+bool CreateFromDump(const std::string& name, const fs::path& wallet_path, bilingual_str& error)
+{
+    // Get the dumpfile
+    std::string dump_filename = gArgs.GetArg("-dumpfile", "");
+    if (dump_filename.empty()) {
+        error = _("No dump file provided. To use createfromdump, -dumpfile=<filename> must be provided.");
+        return false;
+    }
+
+    // Get the data file format
+    std::string file_format = gArgs.GetArg("-format", "");
+    if (file_format.empty()) {
+        error = _("No wallet file format provided. To use createfromdump, -format=<format> must be provided.");
```

Comment:
In commit "wallettool: Add createfromdump command" (fd1f2380cf736df25a27c2607f745ae248c73933)

This seems all right for now, but we should think about how to improve usability in the future because this is inconsistent with other ways wallets are created. Other ways to create wallets either default to BDB format or choose format based on a _descriptors/no descriptors_ option, not a _BDB/SQLite_ format option. Also, dumping and recreating a wallet in the original format may be a challenge here, since both wallet formats look like directories with `wallet.dat` files, and you can't tell them apart without opening them or looking for external clues. This `-format` option could also lead to creation of less well tested BDB+descriptor or SQLite+no-descriptor wallet types.

I think:

- Ideally best format would be detected automatically and there should be no need to have a `-format` option.
- If `-format` option is kept it should give either a warning or an error if an unusual BDB+descriptor or SQLite+no-descriptor wallet is created.
- If `-format` option is kept, wallet `wallet-tool create` command should either respect it or explicitly disallow it, not ignore it. More ideally a `-descriptors`/-`nodescriptors` option would be used to choose the format, so `wallet-tool create` command would be consistent with other ways wallets are created

Reply:
Maybe it should be included in the dump? (Can allow overriding still if user desires to do so)

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19137#discussion_r526240887

> Maybe it should be included in the dump? (Can allow overriding still if user desires to do so)

:+1:

Like that idea

Reply:
Given that this tool is supposed to operate at the database level, I don't think it makes sense to have a `-descriptors`/`-nodescriptors` option for it. Additionally, because this tool doesn't have application level logic, I don't think it should be able to do any introspection of the wallet it creates in order to warn about unusual format and wallet type combinations.

I would prefer that we keep `-format` but make it directly an option of `createfromdump`. It would be better if our args system had something like Python Argparse's subcommands. But I don't think that is currently possible to do with our args system.

Reply:
I've limited `-format` to only `createfromdump`. I've also added a `format` line to the dump which is used to determine the database format if `-format` isn't provided. This required shuffling around some of the code in `createfromdump`. Tests have also been updated and partially refactored to make them a bit more readable.